Apps/Src/Tasks.c

```diff
@@ -45,28 +46,110 @@ CPU_STK DebugDump_Stk[TASK_DEBUG_DUMP_STACK_SIZE];
-        EnterFaultState();
     }
 }
 
+/**
+ * @brief Assert a task error by locking the scheduler (if necessary), displaying a fault screen,
+ * and jumping to the error's specified callback function. 
+ * Called by task-specific error-assertion functions that are also responsible for setting the error variable.
+ * @param errorLoc the task from which the error originated. Note: should be taken out when last task pointer is integrated
+ * @param errorCode the enum for the specific error that happened
+ * @param errorCallback a callback function to a handler for that specific error, 
+ * @param lockSched whether or not to lock the scheduler to ensure the error is handled immediately. Only applicable for recoverable errors- nonrecoverable errors will always lock
+ * @param nonrecoverable whether or not to kill the motor, display the fault screen, and enter an infinite while loop
+*/
+void assertTaskError(os_error_loc_t errorLoc, error_code_t errorCode, callback_t errorCallback, error_scheduler_lock_opt_t lockSched, error_recov_opt_t nonrecoverable) {
+    OS_ERR err;
+
+    if (lockSched == OPT_LOCK_SCHED || nonrecoverable == OPT_NONRECOV) { // Prevent other tasks from interrupting the handling of important (includes all nonrecoverable) errors
+        OSSchedLock(&err);
+        assertOSError(OS_TASKS_LOC, err);
+    }
+
+    // Set the location error variable
+    OSErrLocBitmap = errorLoc; 
+
+    if (nonrecoverable == OPT_NONRECOV) {
+        EmergencyContactorOpen(); // Apart from while loop because killing the motor is more important
+        Display_Error(errorLoc, errorCode); // Needs to happen before callback so that tasks can change the screen
+        // (ex: readCarCAN and evac screen for BPS trip)
+        UpdateDisplay_ClearQueue(); // Clear message queue to ensure no other commands overwrite the error screen
+    }
+
+
+    if (errorCallback != NULL) {
+        errorCallback(); // Run a handler for this error that was specified in another task file
+    }
+    
+
+    if (nonrecoverable == OPT_NONRECOV) { // Enter an infinite while loop
+        while(1) {
+
+            // Print the error that caused this fault
+            printf("\n\rCurrent Error Location: 0x%04x", OSErrLocBitmap);
+            printf("\n\rCurrent Error Code: 0x%04x\n\r", errorCode);
+
+            // Print the errors for each applications with error data
+            printf("\n\rAll application errors:\n\r");
+            printf("Error_ReadCarCAN: 0x%04x\n\r", Error_ReadCarCAN);
+            printf("Error_ReadTritium: 0x%04x\n\r", Error_ReadTritium);
+            printf("Error_UpdateDisplay: 0x%04x\n\r", Error_UpdateDisplay);
+
+            // Delay so that we're not constantly printing
+            for (int i = 0; i < 9999999; i++) {
+            }
+        }
+    }
+
+    if (lockSched == OPT_LOCK_SCHED) { // Only happens on recoverable errors
+        OSSchedUnlock(&err); 
+        // Don't err out if scheduler is still locked because of a timer callback
```

Comment:
The scheduler can have nested locks, and all of the levels must be unlocked before it will run again. However, OSSchedUnlock will still give us an error whenever we unlock a level but have nested locks left. This can happen when we lock and unlock the scheduler inside of a timer callback, ex for ReadCarCAN charge disable messages. Since this is intended behavior, we don't want to assert an OS error.